### PR TITLE
Sync analysis_options.yaml & cleanups

### PR DESCRIPTION
**Pull request by goderbauer**
Syncs the file with the latest list from https://github.com/dart-lang/linter/blob/master/example/all.yaml and enables the following new lints:

* `collection_methods_unrelated_type` (found a real bug: https://github.com/flutter/flutter/issues/117356)
* `enable_null_safety`
* `implicit_call_tearoffs`
* `unnecessary_library_directive`
* `unreachable_from_main` (found a fair amount of dead code)
* `use_string_in_part_of_directives`

To be enabled in a follow-up PR:
* `dangling_library_doc_comments`
* `library_annotations`

### CODE REVIEW

File: dev/bots/test/analyze-snippet-code-test-dart-ui/ui.dart [hunk 1]
```diff
@@ -5,8 +5,7 @@
 // @dart = 2.12
 
 // This is a dummy dart:ui package for the sample code analyzer tests to use.
```
Comment (kevmoo) [inline dev/bots/test/analyze-snippet-code-test-dart-ui/ui.dart:7]: Could just turn this line into a doc comment!

Comment (goderbauer) [reply dev/bots/test/analyze-snippet-code-test-dart-ui/ui.dart:7]: Oh, yeah! I'll do that!

**Review by Piinks — COMMENTED** [review]
CI is sad just now. :(

**Comment by goderbauer** [comment]
Should be fixed now :)

**Comment by goderbauer** [comment]
Never mind, looks like it is still having issues.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Comment by kevmoo** [comment]
This cleanup makes me so happy! 😃

**Pull request merged**